src/boot/efi/pe.c

```diff
@@ -2,177 +2,165 @@
+        #define TARGET_MACHINE_TYPE EFI_IMAGE_MACHINE_IA32
+#elif defined(__x86_64__)
+        #define TARGET_MACHINE_TYPE EFI_IMAGE_MACHINE_X64
+#elif defined(__aarch64__)
+        #define TARGET_MACHINE_TYPE EFI_IMAGE_MACHINE_AARCH64
+#elif defined(__arm__)
+        #define TARGET_MACHINE_TYPE EFI_IMAGE_MACHINE_ARMTHUMB_MIXED
+#elif defined(__riscv) && __riscv_xlen == 64
+        #define TARGET_MACHINE_TYPE EFI_IMAGE_MACHINE_RISCV64
+#else
+        #error Unknown EFI arch
+#endif
+
+static inline BOOLEAN verify_dos(const IMAGE_DOS_HEADER *dos) {
+        assert(dos);
+        return dos->e_magic == IMAGE_DOS_SIGNATURE;
+}
 
-        pe = (struct PeHeader *)&base[dos->ExeHeader];
-        if (CompareMem(pe->Magic, "PE\0\0", 4) != 0)
-                return EFI_LOAD_ERROR;
+static inline BOOLEAN verify_pe(const IMAGE_NT_HEADERS *pe) {
+        assert(pe);
+        return pe->Signature == IMAGE_NT_SIGNATURE &&
+               pe->FileHeader.Machine == TARGET_MACHINE_TYPE &&
+               pe->FileHeader.NumberOfSections > 0 &&
+               pe->FileHeader.NumberOfSections <= MAX_SECTIONS;
+}
 
-        /* PE32+ Subsystem type */
-        if (pe->FileHeader.Machine != PE_HEADER_MACHINE_X64 &&
-            pe->FileHeader.Machine != PE_HEADER_MACHINE_ARM64 &&
-            pe->FileHeader.Machine != PE_HEADER_MACHINE_I386 &&
-            pe->FileHeader.Machine != PE_HEADER_MACHINE_ARM &&
-            pe->FileHeader.Machine != PE_HEADER_MACHINE_RISCV64)
-                return EFI_LOAD_ERROR;
+static inline UINTN section_table_offset(const IMAGE_DOS_HEADER *dos, const IMAGE_NT_HEADERS *pe) {
+        assert(dos);
+        assert(pe);
+        return dos->e_lfanew +
+               OFFSETOF(IMAGE_NT_HEADERS, OptionalHeader) +
+               pe->FileHeader.SizeOfOptionalHeader;
+}
 
-        if (pe->FileHeader.NumberOfSections > 96)
-                return EFI_LOAD_ERROR;
+static VOID locate_sections(
+                const IMAGE_SECTION_HEADER section_table[],
+                UINTN n_table,
+                const CHAR8 **sections,
+                UINTN *addrs,
+                UINTN *offsets,
+                UINTN *sizes) {
 
-        offset = dos->ExeHeader + sizeof(*pe) + pe->FileHeader.SizeOfOptionalHeader;
+        assert(section_table);
+        assert(sections);
+        assert(sizes);
 
-        for (UINTN i = 0; i < pe->FileHeader.NumberOfSections; i++) {
-                struct PeSectionHeader *sect;
+        for (UINTN i = 0; i < n_table; i++) {
+                const IMAGE_SECTION_HEADER *sect = &section_table[i];
 
-                sect = (struct PeSectionHeader *)&base[offset];
                 for (UINTN j = 0; sections[j]; j++) {
-                        if (CompareMem(sect->Name, sections[j], strlena(sections[j])) != 0)
+                        /* If section name length is exactly 8, the name is *not* null terminated!. */
+                        if (strncmpa(sect->Name, sections[j], ELEMENTSOF(sect->Name)) != 0)
```

Comment:
this looks wrong. We know that sections[j] contains NUL terminated strings that are shorter than sect->Name[], hence the old expression was correct and safe: strlena() would always return something <= sizeof(sect->Name[]), because that array is hardcoded in all calls, and if it had longer strings it would be a bad call, i.e a programming error.

If you want to improve the safety of this, maybe add an `assert(strlena(sections[j] <= sizeof(sect->name));`, so that we easily can detect if the function is called with a list of sections with names longer than is allowed.

Or you could do:

```c
if (strlena(sections[j]) > sizeof(sect_name) ||
                CompareMem(sect->name, sections[j], strlen(j]) != 0)
        continue;
```

i.e. filter out overly long sections name and not bother comparing them.

Reply:
You're right, I must've gotten confused at some point while the code was in flux. Turns out the culprit was the bad headerlen computation.